elk/extraction/balanced_sampler.py

```diff
@@ -0,0 +1,99 @@
+from ..utils import infer_label_column
+from ..math_util import stochastic_round_constrained
+from dataclasses import dataclass, field, InitVar
+from datasets import IterableDataset
+from itertools import cycle
+from random import Random
+from torch.utils.data import IterableDataset as TorchIterableDataset
+from typing import Iterator, Optional
+import numpy as np
+
+
+@dataclass
+class BalancedSampler(TorchIterableDataset):
+    """
+    Approximately balances a binary classification dataset in a streaming fashion.
+    Written mostly by GPT-4.
+
+    Args:
+        dataset (IterableDataset): The HuggingFace IterableDataset to balance.
+        label_col (Optional[str], optional): The name of the column containing the
+            binary label. If not provided, the label column will be inferred from
+            the dataset features. Defaults to None.
+        buffer_size (int, optional): The total buffer size to use for balancing the
+            dataset. This value should be divisible by 2, as it will be equally
+            divided between the two binary label values (0 and 1). Defaults to 1000.
+    """
+
+    dataset: IterableDataset
+    label_counts: np.ndarray = field(default_factory=lambda: np.zeros(2))
+    seed: InitVar[int] = 42
+
+    def __post_init__(self, seed: int):
+        self.rng = np.random.default_rng(seed)
+
+    def __iter__(self):
+        for sample in self.dataset:
+            label = sample["label"]
+
+            # Update class counts
+            self.label_counts[label] += 1
+            current_balance = self.label_counts / self.label_counts.sum()
+
+            # Check if the sample should be dropped
+            majority_class = np.argmax(current_balance)
+            if label == majority_class:
+                # Solution of n * p * q / [n * (1 - p) + n * p * q] = 0.5 for q
+                keep_prob = 1 / current_balance[majority_class] - 1
+                if self.rng.uniform() < 1 - keep_prob:
+                    continue
+
+            yield sample
+
+
+class FewShotSampler:
```

Comment:
Any reason why you used @dataclass for the above class, but not here?

Reply:
I haven't had time yet to review it in more detail and to test it. I can do that tomorrow evening, though.

Reply:
yeah this is a little bit inconsistent but idk how much it matters

Reply:
Arguably FewShotSampler might be better if we make it not an iterator bc infinite iterators are weird